File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/service/impl/UserServiceImpl.java
package com.github.arhor.simple.expense.tracker.service.impl;

import lombok.RequiredArgsConstructor;
import lombok.val;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.github.arhor.simple.expense.tracker.data.model.InternalUser;
import com.github.arhor.simple.expense.tracker.data.repository.InternalUserRepository;
import com.github.arhor.simple.expense.tracker.exception.EntityDuplicateException;
import com.github.arhor.simple.expense.tracker.exception.EntityNotFoundException;
import com.github.arhor.simple.expense.tracker.model.Currency;
import com.github.arhor.simple.expense.tracker.model.UserRequestDTO;
import com.github.arhor.simple.expense.tracker.model.UserResponseDTO;
import com.github.arhor.simple.expense.tracker.service.UserService;
import com.github.arhor.simple.expense.tracker.service.mapping.InternalUserMapper;

@Service
@RequiredArgsConstructor(onConstructor_ = @Autowired)
public class UserServiceImpl implements UserService {

    private final InternalUserRepository userRepository;
    private final InternalUserMapper userMapper;

    @Override
    public Long determineUserId(final Authentication auth) {
        val user = determineInternalUser(auth);
        return user.id();
    }

    @Override
    public UserResponseDTO determineUser(Authentication auth) {
        val user = determineInternalUser(auth);
        return userMapper.mapToResponse(user);
    }

    @Override
    @Transactional
    public UserResponseDTO createNewUser(final UserRequestDTO request) {
        val username = request.getUsername();

        if (userRepository.existsByUsername(username)) {
            throw new EntityDuplicateException("InternalUser", "username=" + username);
        }
        val user = userMapper.mapToUser(request);
        val createdUser = userRepository.save(
            user.currency() == null
                ? user.toBuilder().currency("USD").build()
                : user
        );
        return userMapper.mapToResponse(createdUser);
    }

    @Override
    public void createNewUserIfNecessary(final Authentication authentication) {
        if (authentication instanceof OAuth2AuthenticationToken authenticationToken) {
            val externalId = authenticationToken.getName();
            val externalProvider = authenticationToken.getAuthorizedClientRegistrationId();

            if (shouldCreateInternalUser(externalId, externalProvider)) {
                userRepository.save(
                    InternalUser.builder()
                        .externalId(externalId)
                        .externalProvider(externalProvider)
                        .currency(Currency.USD.name())
                        .build()
                );
            }
        }
    }

    private boolean shouldCreateInternalUser(final String externalId, final String externalProvider) {
        return (externalId!=null)
            && (externalProvider!=null)
            && !userRepository.existsByExternalIdAndExternalProvider(externalId, externalProvider);
    }

    private InternalUser.Projection determineInternalUser(final Authentication auth) {
        if (auth instanceof OAuth2AuthenticationToken token) {
            val externalId = token.getName();
            val externalProvider = token.getAuthorizedClientRegistrationId();

            return userRepository.findByExternalIdAndProvider(externalId, externalProvider).orElseThrow(() ->
                new EntityNotFoundException(
                    "User", "externalId=" + externalId + ", externalProvider=" + externalProvider
                )
            );
        }
        if (auth instanceof UsernamePasswordAuthenticationToken token) {
            val username = token.getName();

            return userRepository.findByUsername(username).orElseThrow(() ->
                new EntityNotFoundException(
                    "User", "username=" + username
                )
            );
        }
        throw new IllegalArgumentException(
            "Unsupported authentication type: " + (
                (auth!=null)
                    ? auth.getClass().getSimpleName()
                    :"<null>"
            )
        );
    }
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/service/impl/UserServiceImpl.kt
package com.github.arhor.simple.expense.tracker.service.impl

import com.github.arhor.simple.expense.tracker.data.model.InternalUser
import com.github.arhor.simple.expense.tracker.data.repository.InternalUserRepository
import com.github.arhor.simple.expense.tracker.exception.EntityDuplicateException
import com.github.arhor.simple.expense.tracker.exception.EntityNotFoundException
import com.github.arhor.simple.expense.tracker.model.Currency
import com.github.arhor.simple.expense.tracker.model.UserRequestDTO
import com.github.arhor.simple.expense.tracker.model.UserResponseDTO
import com.github.arhor.simple.expense.tracker.service.CustomUserDetails
import com.github.arhor.simple.expense.tracker.service.UserService
import com.github.arhor.simple.expense.tracker.service.mapping.InternalUserMapper
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserServiceImpl(
    private val userRepository: InternalUserRepository,
    private val userMapper: InternalUserMapper,
) : UserService {

    override fun loadUserByUsername(username: String): CustomUserDetails {
        return userRepository.findInternalUserByUsername(username)?.let(::convertInternalUserToUserDetails)
            ?: throw UsernameNotFoundException("InternalUser with username '$username' is not found.")
    }

    override fun getUserById(userId: Long): UserResponseDTO {
        return userRepository.findByIdOrNull(userId)?.let(userMapper::mapToResponse)
            ?: throw EntityNotFoundException(InternalUser.ENTITY_NAME, "id=${userId}")
    }

    @Transactional
    override fun createNewUser(request: UserRequestDTO): UserResponseDTO {
        val username = request.username

        if (userRepository.existsByUsername(username)) {
            throw EntityDuplicateException(InternalUser.ENTITY_NAME, "username=$username")
        }
        val user = userMapper.mapToUser(request)
        val createdUser = userRepository.save(user)
        return userMapper.mapToResponse(createdUser)
    }

    @Override
    override fun createInternalUserIfNecessary(auth: Authentication) {
        if (auth is OAuth2AuthenticationToken) {
            val externalId = auth.name
            val externalProvider = auth.authorizedClientRegistrationId

            val shouldCreateInternalUser =
                !userRepository.existsByExternalIdAndExternalProvider(
                    externalId,
                    externalProvider
                )

            if (shouldCreateInternalUser) {
                userRepository.save(
                    InternalUser(
                        externalId = externalId,
                        externalProvider = externalProvider,
                        currency = Currency.USD.name
                    )
                )
            }
        }
    }

    // delegate to mapper
    private fun convertInternalUserToUserDetails(it: InternalUser) = CustomUserDetails(
        id = it.id ?: throw IllegalStateException("id cannot be null"),
        currency = it.currency,
        username = it.username ?: throw IllegalStateException("username cannot be null"),
        password = it.password ?: throw IllegalStateException("password cannot be null"),
        authorities = listOf(SimpleGrantedAuthority("ROLE_USER"))
    )
}
